### PR TITLE
iox-#1516 Refactor iceoryx_posh testing lib into own CMake file

### LINKED ISSUE #1516: Support for `gtest_vendor` package
## Brief feature description

When building tests in iceoryx, it checks if a `GTest` package was found and if not it will be build by our own via an ExternalProject in CMake. For a better support on ROS 2, we should add a support for vendor packages for GTest and GMock that are provided by frameworks like ROS 2.

See: https://github.com/ament/googletest/blob/rolling/googletest/package.xml

## Detailed information

- [ ] Try to detect if a gtest vendor package and use this if present instead of the own built GTtest in CMake

### PR DESCRIPTION
**Pull request by dkroenke**
Signed-off-by: Dietrich Kroenke <dietrich.kroenke@apex.ai>

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1516

### CODE REVIEW

File: iceoryx_posh/CMakeLists.txt [hunk 1]
```diff
@@ -299,28 +299,13 @@ setup_install_directories_and_export_package(
...
-                                iceoryx_posh::iceoryx_posh_roudi
-                                iceoryx_hoofs_testing::iceoryx_hoofs_testing
-        FILES
-            testing/roudi_environment/runtime_test_interface.cpp
-            testing/roudi_environment/roudi_environment.cpp
-    )
+# Finding gtest and adding the subdirectories is split to support the use case of
+# building the testing lib without the tests by providing gtest externally
+if(NOT GTest_FOUND AND BUILD_TEST)
+    find_package(GTest CONFIG REQUIRED)
```
Comment (elBoberido) [inline iceoryx_posh/CMakeLists.txt:305]: You need to change this to `if(NOT GTest_FOUND AND (ROUDI_ENVIRONMENT OR BUILD_TEST))` ... we need to implement #1533 :smile:

File: iceoryx_posh/testing/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (elfenpiff) [inline iceoryx_posh/testing/CMakeLists.txt:1]: Please remove this empty line.

File: iceoryx_posh/testing/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
```
Comment (elfenpiff) [inline iceoryx_posh/testing/CMakeLists.txt:2]: Now this becomes weird. I am pretty sure that `iox_add_library` was introduced this year and there is nothing else left in here so why keep the bosch copyright?

But you can keep it if you want ... I find it just more and more silly that we have to drag this everywhere.

**Review by elfenpiff — COMMENTED** [review]
As soon as @elBoberido remarks are handled this is good to merge!

File: iceoryx_posh/CMakeLists.txt (hunk 1 above)
Comment (dkroenke) [reply iceoryx_posh/CMakeLists.txt]: Done

File: iceoryx_posh/CMakeLists.txt [hunk 4]
```diff
@@ -299,28 +299,14 @@ setup_install_directories_and_export_package(
...
-                                iceoryx_posh::iceoryx_posh_roudi
-                                iceoryx_hoofs_testing::iceoryx_hoofs_testing
-        FILES
-            testing/roudi_environment/runtime_test_interface.cpp
-            testing/roudi_environment/roudi_environment.cpp
-    )
+# Finding gtest and adding the subdirectories is split to support the use case of
+# building the testing lib without the tests by providing gtest externally
+if(NOT GTest_FOUND AND (ROUDI_ENVIRONMENT OR BUILD_TEST))
+    find_package(GTest CONFIG REQUIRED)
```
Comment (elBoberido) [inline iceoryx_posh/CMakeLists.txt:305]: Would it be an option to move this into `test/CMakeLists.txt`/`testing/CMakeLists.txt`?

**Review by elBoberido — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

File: .github/workflows/build-test.yml [hunk 5]
```diff
@@ -52,6 +52,7 @@ jobs:
       uses: vmactions/freebsd-vm@v0
       with:
         usesh: true
+        copyback: false
```
Comment (elBoberido) [inline .github/workflows/build-test.yml:55]: Nice. I hope this makes the FreeBSD job more stable

**Review by elBoberido — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**